Added NPM packages support

maicol07 opened the pull request:
This update drops PHP 5.3 support and add the NPM packages support to the library.
JSON extension is required for this to work!

### Installation
In the meantime you can update to my fork adding to `composer.json` this code:
```
"repositories": [
    {
      "type": "vcs",
      "url": "https://github.com/maicol07/Assets"
    }
  ]
```
and modifying this line to:
```
"stolz/assets": "dev-master"
```

maicol07 committed:
Added NPM packages support

maicol07 committed:
Added docs for NPM integration

maicol07 committed:
fixup! Added docs for NPM integration

maicol07 committed:
Search for JS using different props

maicol07 committed:
Changed fork ownership from @stolz to @maicol07

maicol07 committed:
Composer package adjustements

maicol07 committed:
Composer package adjustements

maicol07 committed:
Fix

maicol07 committed:
Fix CSS of NPM packages

maicol07 committed:
Fix CSS of NPM packages

maicol07 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/Manager.php

maicol07 committed:
Removed 5.3 in Travis CI builds

Stolz commented:
Hi @maicol07,

First apologies for my late reply. Second I appreciate your initiative but I'm afraid this is not going to get merged for several reasons:

- It modifies almost every line of code of the main library for no reason, masquerading the original contributors. There are also other netiquette related issues such adding personal branding to the source code of a Git based repo that I think I don't need to explain. 
- Drops support for older versions of PHP. I'm in favor in doing so as well as in modernizing the code style (to PSR-2 or even PSR-12) but this should be done as a standalone step, not as part of a new feature.
- It requires JSON extension in `composer.json`. This prevents users who don't use/need NPM from installing the library if they don't have the extension. Lack of the extension should be handled in run time only when the NPM feature is used.
- It assumes the library is installed in your public dir (`$docroot = __DIR__`) which will never be true when installed via Composer.
- It seems out of scope for a library that focuses itself in simplicity and being framework agnostic to include support for every possible assets package manager out there. I know this PR is only for one package manager (NPM/Yarn) but think of the implications if this is added, it will leave the door open for others to follow the same path therefore deviating from the philosophy of the library. Besides, the library already has support for a generic package loading feature as well as a a feature to load assets from a directory which with minor tweaks could possibly be leveraged to achieve the same effect.
- The new feature is not documented. You updated the `API.md` file (which BTW when updated via `bin/generate_doc` script yields a different file than your one) but there is no reference of the feature in the README file, making it virtually invisible.
- The new feature lacks automatic tests.
- Code quality and excessive complexity IMHO are not worth for the added value. The Cyclomatic and NPath complexities of the new code are unhealthily high (the highest of the whole library in fact). In other words, the NPM feature is the most complex part of the library.

Let me know your thoughts.

maicol07 commented:
Hi @Stolz,
thanks for your reply. I'll answer you point by point:
- I don't think my PR edits so much the basic structure of the library (Github shows that almost all is modified because my IDE (Phpstorm) has optimized indents (i. e. 4 spaces = 1 tab)). I've only added the NPM/Yarn feature with the required edits in other methods to work properly.
- PHP 5.3 is indeed an old and almost unused version of PHP, that lacks security and performance (compared to newest versions like PHP 7.x). You can do this as a separate update and then I'll merge your changes to my fork.
- I agree with you (actually, when I wrote the function I haven't thought about this) so I've added the check for the JSON extension when searching for the NPM package and moved `ext-json` to `suggest` section of `composer.json`
- I've also deleted the default `$docroot` value. This requires adding it to config if you want to use the NPM/Yarn package search.
- I know this, but NPM/Yarn is the most used package manager around the web (actually, I don't know others except for Bower but is outdated and deprecated by developers). Almost every CSS/JS lib is on the NPM package registry.
- I didn't know about that script... However, this scripts returns an error:
```
Collecting files .. OK
Initializing parser .. OK
Parsing files
Parsing D:\Maicol\Documents\Progetti\Web\Assets\src\Manager.php
  No summary was found for this file
  No summary for class \Stolz\Assets\Manager
Storing cache in "D:\Maicol\Documents\Progetti\Web\Assets\doc" .. OK
Load cache                                                         ..    0.196s
Preparing template "xml"                                           ..    0.009s
Preparing 1 transformations                                        ..    0.000s
Build "elements" index                                             ..    0.000s
Replace textual FQCNs with object aliases                          ..    0.001s
Resolve @link and @see tags in descriptions                        ..    0.000s
Enriches inline example tags with their sources                    ..    0.000s
Build "packages" index                                             ..    0.053s
Build "namespaces" index and add namespaces to "elements"          ..    0.000s
Collect all markers embedded in tags                               ..    0.000s
Transform analyzed project into artifacts                          ..
Applying 1 transformations
  Initialize writer "phpDocumentor\Plugin\Core\Transformer\Writer\Xml"
  Execute transformation using writer "xml"
   0.099s
Analyze results and write report to log                            ..    0.000s
Parsing structure.xml
Generating pages
PHP Notice:  Array to string conversion in D:\Maicol\Documents\Progetti\Web\Assets\vendor\twig\twig\src\Loader\FilesystemLoader.php on line 39

Notice: Array to string conversion in D:\Maicol\Documents\Progetti\Web\Assets\vendor\twig\twig\src\Loader\FilesystemLoader.php on line 39
PHP Warning:  realpath() expects parameter 1 to be a valid path, array given in D:\Maicol\Documents\Progetti\Web\Assets\vendor\twig\twig\src\Loader\FilesystemLoader.php on line 40

Warning: realpath() expects parameter 1 to be a valid path, array given in D:\Maicol\Documents\Progetti\Web\Assets\vendor\twig\twig\src\Loader\FilesystemLoader.php on line 40
Complete
```
- I don't know how to do automatic tests... (I've never used PHPUnit...)
- I know it's complex but there isn't a standard value in `package.json`... every developer decides their own value, so the function searchs for the most commons...

I'll continue mantaining the fork so everyone who wants to use this feature can download it. However, with your help maybe we can simplify the most complex things...
Thank you for your understanding
